### PR TITLE
Support [[guarded_by(mutex)]] attribute inside C struct

### PR DESCRIPTION
**Pull request by pdherbemont**
Today, it's only supported inside C++ classes or top level C/C++
declaration.

I mostly copied and adapted over the code from the
[[counted_by(value)]] lookup.

### CODE REVIEW

File: clang/lib/Parse/ParseDecl.cpp
Comment (pdherbemont) [inline clang/lib/Parse/ParseDecl.cpp:3369]: I am looking for guidance here: should I introduce a new `ExpressionKind` for `GuardedByAttr`, or simply rename it?

File: clang/lib/Parse/ParseDecl.cpp
Comment (AaronBallman) [inline clang/lib/Parse/ParseDecl.cpp:683]:
```suggestion
  } else if (AttrKind == ParsedAttr::AT_GuardedBy ||
             AttrKind == ParsedAttr::AT_PtGuardedBy) {
    ParseGuardedByAttribute(*AttrName, AttrNameLoc, Attrs, ScopeName, ScopeLoc,
                            EndLoc,
                            Form);
    return;
```

File: clang/lib/Parse/ParseDecl.cpp
Comment (AaronBallman) [reply clang/lib/Parse/ParseDecl.cpp:3369]: It seems like this should be renamed to `EK_AttrArgument` as it applies more generally to attribute arguments; WDYT @erichkeane @rapidsna ?

File: clang/lib/Parse/ParseDecl.cpp
Comment (AaronBallman) [inline clang/lib/Parse/ParseDecl.cpp:3388]:
```suggestion
  if (EndLoc)
     *EndLoc = Tok.getLocation();
```

**Review by AaronBallman — COMMENTED** [review]
Thank you for working on this!

The changes should come with a release note in clang/docs/ReleaseNotes.rst so users know about the new functionality.

One question I have is whether there will be follow-up work for other thread safety attributes (`acquire_capability`, `try_acquire_capability`, etc)?

File: clang/lib/Parse/ParseDecl.cpp
Comment (rapidsna) [reply clang/lib/Parse/ParseDecl.cpp:3369]: @AaronBallman Renaming it makes a lot of sense to me.

**Review by rapidsna — COMMENTED** [review]
You may also want to consider making the attribute late parsed in C when `-fexperimental-late-parse-attributes` is enabled. See https://github.com/llvm/llvm-project/pull/93121/files#diff-ae2ec9524bdbeea1f06917607482634dd89af5bcbb929805032463e5dafe79e7R2260

That will allow the code like below:
```
 struct Foo {
   int a_value GUARDED_BY(mu_); // attribute comes before `mu_` which needs to be late parsed
   struct Mutext *mu_;
 }
```

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (rapidsna) [inline clang/test/Sema/warn-thread-safety-analysis.c:35]: Nit: Please check indent

**Comment by pdherbemont** [comment]
> One question I have is whether there will be follow-up work for other thread safety attributes (`acquire_capability`, `try_acquire_capability`, etc)?

They already work fine in C so far. But if you are aware of some issues don't hesitate to let me know!

File: clang/lib/Parse/ParseDecl.cpp
Comment (pdherbemont) [reply clang/lib/Parse/ParseDecl.cpp]: Switched to `EK_AttrArgument` and fixed the style (using the clang-format) thank you for the feedback!

File: clang/lib/Parse/ParseDecl.cpp
Comment (pdherbemont) [reply clang/lib/Parse/ParseDecl.cpp]: Done!

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (pdherbemont) [reply clang/test/Sema/warn-thread-safety-analysis.c:35]: Fixed

**Review by Endilll — COMMENTED** [review]
`Sema.h` changes look good to me.

**Comment by AaronBallman** [comment]
> > One question I have is whether there will be follow-up work for other thread safety attributes (`acquire_capability`, `try_acquire_capability`, etc)?
> 
> They already work fine in C so far. But if you are aware of some issues don't hesitate to let me know!

Ah, sorry, of course I picked examples that appertain to functions rather than fields. But some of the other thread safety attributes have the same problem as `guarded_by`: https://godbolt.org/z/hWTMxzrns

**Comment by pdherbemont** [comment]
> You may also want to consider making the attribute late parsed in C when `-fexperimental-late-parse-attributes` is enabled. See https://github.com/llvm/llvm-project/pull/93121/files#diff-ae2ec9524bdbeea1f06917607482634dd89af5bcbb929805032463e5dafe79e7R2260
> 
> That will allow the code like below:
> 
> ```
>  struct Foo {
>    int a_value GUARDED_BY(mu_); // attribute comes before `mu_` which needs to be late parsed
>    struct Mutext *mu_;
>  }
> ```

Adopted `LateAttrParseExperimentalExt`. Let me know if that looks okay.

**Comment by pdherbemont** [comment]
> > > One question I have is whether there will be follow-up work for other thread safety attributes (`acquire_capability`, `try_acquire_capability`, etc)?
> > 
> > 
> > They already work fine in C so far. But if you are aware of some issues don't hesitate to let me know!
> 
> Ah, sorry, of course I picked examples that appertain to functions rather than fields. But some of the other thread safety attributes have the same problem as `guarded_by`: https://godbolt.org/z/hWTMxzrns

Oh yes – maybe it makes sense to fix them in this commit. Let me try to look into that quickly.

**Comment by pdherbemont** [comment]
> > > > One question I have is whether there will be follow-up work for other thread safety attributes (`acquire_capability`, `try_acquire_capability`, etc)?
> > > 
> > > 
> > > They already work fine in C so far. But if you are aware of some issues don't hesitate to let me know!
> > 
> > 
> > Ah, sorry, of course I picked examples that appertain to functions rather than fields. But some of the other thread safety attributes have the same problem as `guarded_by`: https://godbolt.org/z/hWTMxzrns
> 
> Oh yes – maybe it makes sense to fix them in this commit. Let me try to look into that quickly.

Seems like it's an easy change for `acquired_after` and `acquired_before`! Added support & a test.

File: clang/lib/Parse/ParseDecl.cpp
Comment (pdherbemont) [reply clang/lib/Parse/ParseDecl.cpp]: Renamed!

**Comment by pdherbemont** [comment]
Pushed an additional change on the documentation: the doc was incorrectly stating that ACQUIRED_{AFTER,BEFORE} were not implemented but they are.

**Comment by rapidsna** [comment]
> > You may also want to consider making the attribute late parsed in C when `-fexperimental-late-parse-attributes` is enabled. See https://github.com/llvm/llvm-project/pull/93121/files#diff-ae2ec9524bdbeea1f06917607482634dd89af5bcbb929805032463e5dafe79e7R2260
> > That will allow the code like below:
> > ```
> >  struct Foo {
> >    int a_value GUARDED_BY(mu_); // attribute comes before `mu_` which needs to be late parsed
> >    struct Mutext *mu_;
> >  }
> > ```
> 
> Adopted `LateAttrParseExperimentalExt`. Let me know if that looks okay.

Thanks! Could you please also add a test taking advantage of late parsing?

**Comment by pdherbemont** [comment]
> `-fexperimental-late-parse-attributes`

Done

> > > You may also want to consider making the attribute late parsed in C when `-fexperimental-late-parse-attributes` is enabled. See https://github.com/llvm/llvm-project/pull/93121/files#diff-ae2ec9524bdbeea1f06917607482634dd89af5bcbb929805032463e5dafe79e7R2260
> > > That will allow the code like below:
> > > ```
> > >  struct Foo {
> > >    int a_value GUARDED_BY(mu_); // attribute comes before `mu_` which needs to be late parsed
> > >    struct Mutext *mu_;
> > >  }
> > > ```
> > 
> > 
> > Adopted `LateAttrParseExperimentalExt`. Let me know if that looks okay.
> 
> Thanks! Could you please also add a test taking advantage of late parsing?

Done!

**Review by rapidsna — APPROVED** [review]
LGTM assuming you will be fixing the buildkite failures. 

You may still need to wait on @AaronBallman's approval.

**Comment by pdherbemont** [comment]
> LGTM assuming you will be fixing the buildkite failures.
> 
> You may still need to wait on @AaronBallman's approval.

Fixed the buildkite issue – they were regression from the latest changes about acquired_{before,after} handling. Also pinged @AaronBallman. Thank you all for the feedback!

**Review by AaronBallman — APPROVED** [review]
LGTM! Do you need someone to land these changes on your behalf?

File: clang/lib/Parse/ParseDecl.cpp
Comment (delcypher) [inline clang/lib/Parse/ParseDecl.cpp:3387]: @pdherbemont Could you add a test case for this path?

File: clang/lib/Parse/ParseDecl.cpp
Comment (delcypher) [inline clang/lib/Parse/ParseDecl.cpp:3355]: @pdherbemont Could you add a test case for this path?

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (delcypher) [inline clang/test/Sema/warn-thread-safety-analysis.c:1]: @pdherbemont We should probably have a test case that runs without `-fexperimental-late-parse-attributes` so that we check normal parsing still works, especially given that its the default.

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (delcypher) [inline clang/test/Sema/warn-thread-safety-analysis.c:32]: @pdherbemont I think you should be checking **both** late and regular parsing for all attributes that you're adding support for.

File: clang/lib/Parse/ParseDecl.cpp
Comment (delcypher) [inline clang/lib/Parse/ParseDecl.cpp:3372]: @pdherbemont Could you add a test case for this path?

File: clang/lib/Parse/ParseDecl.cpp
Comment (delcypher) [inline clang/lib/Parse/ParseDecl.cpp:3409]: @pdherbemont Could you add a test case for this path?

File: clang/lib/Parse/ParseDecl.cpp
Comment (delcypher) [inline clang/lib/Parse/ParseDecl.cpp:3430]: @pdherbemont Could you add a test case for this path?

**Review by delcypher — CHANGES_REQUESTED** [review]
@pdherbemont thanks for working on this. It looks pretty good and its great to see the late parsing support I added gaining new users. I just have some nits about missing test cases.

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (pdherbemont) [reply clang/test/Sema/warn-thread-safety-analysis.c]: done!

File: clang/lib/Parse/ParseDecl.cpp
Comment (pdherbemont) [reply clang/lib/Parse/ParseDecl.cpp:3355]: done!

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (pdherbemont) [reply clang/test/Sema/warn-thread-safety-analysis.c:33]: done!

File: clang/lib/Parse/ParseDecl.cpp
Comment (pdherbemont) [reply clang/lib/Parse/ParseDecl.cpp:3409]: done!

File: clang/lib/Parse/ParseDecl.cpp
Comment (pdherbemont) [reply clang/lib/Parse/ParseDecl.cpp:3372]: done!

File: clang/lib/Parse/ParseDecl.cpp
Comment (pdherbemont) [reply clang/lib/Parse/ParseDecl.cpp:3430]: done!

**Comment by pdherbemont** [comment]
> @pdherbemont thanks for working on this. It looks pretty good and its great to see the late parsing support I added gaining new users. I just have some nits about missing test cases.

I should have added all the requested test cases. Not sure it's the way you wanted it, so let me know if you want any change!

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (delcypher) [inline clang/test/Sema/warn-thread-safety-analysis.c:91]: @pdherbemont I think you're missing a use of `pt_guarded_by` and `acquired_before` being late parsed here.

Comment (delcypher) [reply clang/test/Sema/warn-thread-safety-analysis.c:91]: @pdherbemont Also I can't see any uses of the `LateParsing` struct. It might be worth testing the attributes were parsed. You could do that as a separate AST test or use the `LateParsing` struct in a way that will make the thread safety analysis emit warnings.

**Review by delcypher — CHANGES_REQUESTED** [review]

**Review by delcypher — APPROVED** [review]
Thanks for fixing my requested changes. LGTM

**Comment by delcypher** [comment]
@pdherbemont I've landed on your behalf. I tweaked the commit message to be more descriptive of the change we finally landed and noted that you are the author of this patch. Thanks again for working on this :)

**Comment by aeubanks** [comment]
this seems to have broken code like [this](https://crsrc.org/c/base/allocator/partition_allocator/src/partition_alloc/random.cc;drc=36293863bf9bbc8131797eb8f4d2bafe8af3de79;l=33) with
```
../../base/allocator/partition_allocator/src/partition_alloc/random.cc:33:69: error: invalid use of non-static data member 'lock_'
   33 |     internal::base::InsecureRandomGenerator instance_ PA_GUARDED_BY(lock_);
      |                                                                     ^~~~~
../../base/allocator/partition_allocator/src/partition_alloc/partition_alloc_base/thread_annotations.h:59:70: note: expanded from macro 'PA_GUARDED_BY'
   59 | #define PA_GUARDED_BY(x) PA_THREAD_ANNOTATION_ATTRIBUTE__(guarded_by(x))
      |                                                                      ^
../../base/allocator/partition_allocator/src/partition_alloc/partition_alloc_base/thread_annotations.h:44:60: note: expanded from macro 'PA_THREAD_ANNOTATION_ATTRIBUTE__'
   44 | #define PA_THREAD_ANNOTATION_ATTRIBUTE__(x) __attribute__((x))
      |                                                            ^
../../base/allocator/partition_allocator/src/partition_alloc/random.cc:35:65: error: invalid use of non-static data member 'lock_'
   35 |         internal::base::InsecureRandomGenerator)] PA_GUARDED_BY(lock_) = {};
      |                                                                 ^~~~~
../../base/allocator/partition_allocator/src/partition_alloc/partition_alloc_base/thread_annotations.h:59:70: note: expanded from macro 'PA_GUARDED_BY'
   59 | #define PA_GUARDED_BY(x) PA_THREAD_ANNOTATION_ATTRIBUTE__(guarded_by(x))
      |                                                                      ^
../../base/allocator/partition_allocator/src/partition_alloc/partition_alloc_base/thread_annotations.h:44:60: note: expanded from macro 'PA_THREAD_ANNOTATION_ATTRIBUTE__'
   44 | #define PA_THREAD_ANNOTATION_ATTRIBUTE__(x) __attribute__((x))
      |                                                            ^
```

is that intentional?

**Comment by aeubanks** [comment]
reduced:
```
struct Lock {};
struct A {
        Lock lock;
        union {
                bool b __attribute__((guarded_by(lock)));
        };
};
```

seems like a regression, can we revert for now?

**Comment by delcypher** [comment]
@aeubanks I'll revert. Is this example C or C++?

**Comment by delcypher** [comment]
Reverted in c9d580033f29196223cd56a0fa238c3ba51d23e4

**Comment by delcypher** [comment]
@pdherbemont Could you take a look at this?

**Comment by aeubanks** [comment]
> @aeubanks I'll revert. Is this example C or C++?

C++

**Comment by pdherbemont** [comment]
Ah nice catch! Thank you @aeubanks and @delcypher – I am going to try to understand why (and add a new test)

**Comment by pdherbemont** [comment]
So I found the cause of the regression: We used to parse the argument in an `Unevaluated` context in C++ mode. We now parse them in a `PotentiallyEvaluated` context. Switching to `Unevaluated` makes the C parsing code fail. A simple fix is to switch the context based on the language detected.

**Comment by pdherbemont** [comment]
Follow-up PR #95455

**Comment by bvanassche** [comment]
Thread-safety attributes of cleanup functions are ignored. Does anyone know how to fix this? See also https://github.com/llvm/llvm-project/issues/122758.

**Pull request merged**
